### DISCUSSION TITLE
Query ORM relationships with AsyncSession

### DISCUSSION THREAD

**Discussion by mockingjet** [sqlalchemy/sqlalchemy#6918] [Usage Questions] [answered]
```py
class UserModel(Base):
    __tablename__ = "users"

    id = sa.Column(sa.String, primary_key=True)
    username = a.Column(sa.String, nullable=False)

    exams =  sa.orm.relationship("ExamModel")
   
    # I have to judge if I query this user with books selectinload, then I can use this method
    # It's easy to call this method without books selecinload, and emit async/await errors
    def get_total_scores(self):
        return sum([exam.score for exam in self.exams)

    # if not with books selectinload I have to use async method
    async def get_total_scores(self, async_session):
        me = await async_session.get(
            UserModel,
            self.id,
            [sa.orm.selectinload(UserModel.exams)]
        )
        
        return sum([exam.score for exam in me.exams)

class ExamModel(Base):
    __tablename__ = "books"

    id = sa.Column(sa.Integer, primary_key=True)
    score = sa.Column(sa.String, nullable=False)

    user_id = sa.Column(sa.ForeignKey("users.id"))
```

How to retrieve user's books in model layer with AsyncSession?

I provide an example I came up with above. And every time I want to get a user's total scores, I need to pass async_session as argument first and user it to query user himself with selectinload. It seems not straightforward compared to sync method. So I am here looking for a simpler implementation. 

Similar to this [discussion](https://github.com/sqlalchemy/sqlalchemy/discussions/6783)

Perhaps my async / await concept is totally wrong, then please correct me, thanks.

**Comment by zzzeek** [comment, marked as answer]
Well this comes down to the whole philosophy of asyncio, which is, "never run any IO implicitly".

if you're using asyncio, that means you are looking for this pattern - you dont want to load any data from the database without being explicit.  So that you can't just say user.exams without using "await" is the reason people want to use asyncio.

I would say if you *dont* like that, then you'd stick to sync.   that said, yes you use eager loading like selectinload up front to make sure you have the data you need.   

if you want to break the rules a little bit, then there is the [run_sync](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#running-synchronous-methods-and-functions-under-asyncio) approach.   here you could get "user.exams" either in a fully sync-style function for the whole thing you want to do as in the linked example, or, you could run the lazyload explicitly like this:

     exams = await session.run_sync(lambda s: user.exams)

which is pretty much what I said in that other discussion.   

I'm not sure what other approach you'd want to see ?    how do you think it should look ?

**Reply by mockingjet** [reply]
> you can't just say user.exams without using "await" is the reason people want to use asyncio

Yes, I know this philosophy from that discussion. Maybe I just can't get used to it because I used sync before and it's too straightforward.

Now using AsyncSession makes me afraid to use relationships-related methods in model layer, since I can not assure that instance have already fetched children or parent. It seems better to move all model methods to service or repository layer. And in this case it's simpler to use sqlalchemy.Table instead of sqlalchemy.orm.mapper

**Comment by CaselIT** [comment]
If the methods are only called with objects attached to a session, you can use the [object_session](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.object_session).
